Separate hive integration test S3 prefixes

sameer2086 opened the pull request:
This change fixes flaky TestHiveFileSystemS3SelectPushdown and
TestHiveFileSystemS3 integration tests by separating the data
set S3 prefixes for the two tests. The tests are correct only when
the data set is idempotent, as concurrent test runs will try to
update the same location.

https://github.com/prestodb/presto/issues/12153

findepi commented on presto-hive-hadoop2/bin/run_hive_s3_tests.sh:
```diff
@@ -24,6 +24,14 @@ exec_in_hadoop_master_container hadoop fs -copyFromLocal -f /tmp/test1.csv.lz4 "
 exec_in_hadoop_master_container hadoop fs -copyFromLocal -f /tmp/test1.csv.bz2 "${table_path}"
```
I think to actually fix https://github.com/prestodb/presto/issues/12153, you would need revert the addition of `/tmp/test1.csv.*` here, into the `s3a://${S3_BUCKET}/presto_test_external_fs/` location

sameer2086 replied:
Additional `test1.csv.*` files under `s3a://${S3_BUCKET}/presto_test_external_fs/` represent additional duplicate rows in `presto_test_external_fs` table. As all of the tests pass locally, I believe, they shouldn't be asserting on the exact content of the file, except for `testGetRecords` that we have already updated. 

Ideally, a table could consist of multiple objects and thus the tests should be resilient of the change of adding a new supported object. Also, the end-to-end happy case tests that are supported by S3 SELECT should always be supported by S3 GET.  Am I missing something over here? Should we fix the specific tests that couldn't handle multiple files instead?

sameer2086 replied:
@electrum, @findepi, @kokosing Do we want to address `%RANDOM_PART%` as a part of this PR or should we create an issue instead and close this PR?

kokosing replied:
I would close this one and create another issue and PR.

sameer2086 replied:
Sure will close this PR then.

findepi replied:
I would not create another issue. 
https://github.com/prestodb/presto/issues/12153 got addressed internally and closed, and i don't see a need for more changes currently.

sameer2086 commented:
Incorporated @electrum's change request and rebased with master. Awaiting @findepi feedback on https://github.com/prestodb/presto/pull/12157#discussion_r244655988.

findepi commented:
@sameer2086 what does this PR actually fix?

maybe i should explain https://github.com/prestodb/presto/issues/12153 a little bit better, sorry for not doing this earlier.

we run `TestHiveFileSystemS3` internally against some test bucket. 
When we run the test on master, setup script uploads several files.
When we run it on a stable release branch, setup script would upload just one file (which is already there). The test will expect data from one file (`testGetRecords` test), so it will fail, as Presto will (correctly) read the data from all the files found.

Perhaps this is a fault on our side -- we could use different buckets. Using single bucket was just opportunistically simplest, but maybe not correct.
If we do a change on our side, then I don't see a need to merge this PR.

sameer2086 commented:
> @sameer2086 what does this PR actually fix?

The PR separates the artifacts for the two tests. I don't have complete understanding of Hive S3 integration tests and different configurations in which the tests are executed. I believed it would be best to separate the artifacts for the two tests to avoid any mutations (in current or future tests) that might lead to one test overwriting mutations of the other. I should have dived deeper to understand your use case better. From the digging I did, my understanding was that all mutations are done in random tmp prefixes which should be safe. However, this PR seemed as an additional value add regardless.

>we run TestHiveFileSystemS3 internally against some test bucket.
When we run the test on master, setup script uploads several files.
When we run it on a stable release branch, setup script would upload just one file (which is already there). The test will expect data from one file (testGetRecords test), so it will fail, as Presto will (correctly) read the data from all the files found.

Thanks for sharing the details. Now the problem makes more sense. Why does the stable release not use master's test setup?

>Perhaps this is a fault on our side -- we could use different buckets. Using single bucket was just opportunistically simplest, but maybe not correct.

Rather that using a single test1.csv, could the stable release tests be updated to support the same number of files as master? 

>If we do a change on our side, then I don't see a need to merge this PR.

I still think that this PR would add value as then we could write separate tests for different tests for S3SelectPushdown vs S3 GET in future. I am ok with updating the commit to "**Use separate S3 prefixes for integration tests**"

findepi commented:
> two tests to avoid any mutations (in current or future tests) that might lead to one test overwriting mutations of the other

The data cannot be mutated by the test. The data must be uploaded in an idempotent manner and stay umodified.
If we were to support data mutations, we would need to
1. write to a random location within the bucket and make sure we clean up afterewards (best to do this test Java code; doing cleanups in shell script is cumbersome, since bash has equivalent of `finally`)
2. OR use fresh bucket each time we run tests

Since the second option will make harder to anyone else to run the tests, i strongly prefer the first option.

> Rather that using a single test1.csv, could the stable release tests be updated to support the same number of files as master?

Thanks, this is what we will do. I am closing https://github.com/prestodb/presto/issues/12153

sameer2086 commented:
> I would close this one and create another issue and PR.

Closing the PR as per @kokosing's response. The work to make Presto Hive S3 integration tests more robust, by separating S3 prefixes, will be done in a separate PR and issue.
